store-encryption: Use bincode for encoding

DarkKirb opened the pull request:
This commit changes the default ciphertext serialization format for matrix-sdk-store-encryption to bincode, as the current json serialization is incurring a large space overhead.

Currently, the following plaintext encrypts to ~237 bytes. With this change, it only takes 58 bytes (75% reduction in size):

```
{"test":"test"}
```

Existing ciphertexts are unaffected, and still supported

jplatte commented on crates/matrix-sdk-store-encryption/src/lib.rs:
```diff
@@ -54,6 +67,36 @@ pub enum Error {
+#[cfg(not(feature = "bincode"))]
+fn deserialize<T>(value: &[u8]) -> Result<T, SerializationError>
+where
+    T: for<'de> Deserialize<'de>,
```
```suggestion
    T: DeserializeOwned,
```

DarkKirb replied:
The other instances of Deserialize trait bounds in the file also don’t use DeserializeOwned, but i can change that in another   commit

jplatte replied:
I posted a PR for that earlier: #687

DarkKirb replied:
I guess I duplicated the changes here then

jplatte replied:
Yeah, but git should merge that just fine.

DarkKirb commented:
`client::tests::forget_room` seems flakey, can’t reproduce the failure locally

gnunicorn commented on crates/matrix-sdk-store-encryption/Cargo.toml:
```diff
@@ -9,8 +9,10 @@ rust-version = "1.60"
 
 [features]
 js = ["getrandom/js"]
+default = ["bincode"]
```
so, wait. this would include both bincode and json_serde then?

Also, if we changed this default, we'd have a breaking change here - as it won't be able deserialise value written before with serde_json. I can see two possible ways out:
1. make bincode non-default and disable serde_json when not in use (which is what I'd prefer for now) on this crate
2. provide fallback support, to read old values with serde_json, too - but only encrypt on save with the new one. Not sure how feasible this is though...
Either way, for now:

```suggestion
default = []
bincode = ["dep:bincode"]
```

DarkKirb replied:
the deserialize helper function attempts to read with serde_json first, then bincode. I added a test that makes sure that old ciphertexts could still be decoded (`decrypt_json`)

gnunicorn reviewed (changes requested):
the way this is implemented right now  would break current databases out there without the authors knowing about that or any way to migrate that. Either we make that a non-default option (so it must actively be switched on) or we provide migration support...

poljar commented on crates/matrix-sdk-store-encryption/src/lib.rs:
```diff
@@ -39,11 +39,24 @@ const KDF_ROUNDS: u32 = 1000;
+pub enum SerializationError {
+    /// Failed to serialize or deserialize a JSON value {0}
+    Json(#[from] serde_json::Error),
+    #[cfg(feature = "bincode")]
```
Docs go above attributes.

poljar commented on crates/matrix-sdk-store-encryption/src/lib.rs:
```diff
@@ -39,11 +39,24 @@ const KDF_ROUNDS: u32 = 1000;
+    #[cfg(feature = "bincode")]
+    /// Failed to serialize a bincode value {0}
+    BincodeSerialization(#[from] bincode::error::EncodeError),
+    #[cfg(feature = "bincode")]
```
Same.

poljar reviewed (commented):
So this avoids the inefficiency by side-stepping it. The problem still exists for `serde_json` users. #691 mitigates this mostly.

I still don't think we should add more encoders to this crate. People can use any serde based encoder.

 If people don't want to use the serde based one, we can make the fields of `EncryptedValue` public so they can side-step this by using the `decrypt_value_data()` and `encrypt_value_data()` methods. They can then transform the `EncryptedValue` into any other type.

DarkKirb commented:
On 19.05.22 08:35, Benjamin Kampmann wrote:
> @gnunicorn requested changes on this pull request.
> 
> the way this is implemented right now  would break current databases out there without the authors knowing about that or any way to migrate that. Either we make that a non-default option (so it must actively be switched on) or we provide migration support...
> 
>> @@ -9,8 +9,10 @@ rust-version = "1.60"
>   
>   [features]
>   js = ["getrandom/js"]
> +default = ["bincode"]
> 
> so, wait. this would include both bincode and json_serde then?
> 
> Also, if we changed this default, we'd have a breaking change here - as it won't be able deserialise value written before with serde_json. I can see two possible ways out:
> 1. make bincode non-default and disable serde_json when not in use (which is what I'd prefer for now) on this crate

the deserialize helper function attempts to read with serde_json first, 
then bincode. I added a test that makes sure that old ciphertexts could 
still be decoded (`decrypt_json`)

> 2. provide fallback support, to read old values with serde_json, too - but only encrypt on save with the new one. Not sure how feasible this is though...
> Either way, for now:
> 
> ```suggestion
> default = []
> bincode = ["dep:bincode"]
> ```
>

DarkKirb commented:
> So this avoids the inefficiency by side-stepping it

I don’t think there is any reason for the data to be json-encoded at all. When I started using it for the matrix-sdk-sql crate, I was under the impression it was encoded with a binary format, similar to what this PR adds.
Base64 encoding like in #691 would still add a 33% size overhead over returning binary data.
The reason why i enabled bincode by default originally is because the patch would allows for a smooth migration to the binary encoding.

DarkKirb commented on crates/matrix-sdk-store-encryption/Cargo.toml:
```diff
@@ -9,8 +9,10 @@ rust-version = "1.60"
+default = ["bincode"]
 
 [dependencies]
+bincode = { version = "2.0.0-rc.1", optional = true }
```
Switched to 1.3.3

poljar commented:
> I don’t think there is any reason for the data to be json-encoded at all. When I started using it for the matrix-sdk-sql crate, I was under the impression it was encoded with a binary format, similar to what this PR adds. Base64 encoding like in #691 would still add a 33% size overhead over returning binary data. The reason why i enabled bincode by default originally is because the patch allows for a smooth migration to the binary encoding.

Since you remain unconvinced, perhaps we should just remove the json based one and force users to pick one themselves. I was under the impression that bincode 2.0 removes serde support, but the [docs](https://docs.rs/bincode/2.0.0-rc.1/bincode/serde/index.html) seem to suggest otherwise.

gnunicorn commented:
@DarkKirb , what do you think about @poljar's proposal?

DarkKirb commented:
I ended up doing that in matrix-sdk-sql

gnunicorn commented:
@DarkKirb does this mean this PR should be closed un-merged then?

DarkKirb commented:
that’s up to you to decide, I still think that putting the change in the store-encryption crate is a cleaner solution

poljar commented:
I'm not really a fan of bincode, I tried to use it a couple of times and hit some issues so I would prefer no default over a bincode default.

Closing this and we'll remove the default format as part of https://github.com/matrix-org/matrix-rust-sdk/issues/718.

DarkKirb commented:
On 27.05.22 09:29, Benjamin Kampmann wrote:
> @DarkKirb , what do you think about @poljar's proposal?
> 
I ended up doing that in matrix-sdk-sql
